### PR TITLE
Enable creating and inserting to empty external tables via SQL

### LINKED ISSUE #7228: Create new empty external table
### Is your feature request related to a problem or challenge?

Thanks to the great work from @metesynnada @devinjdangelo  and others, it is now possible to have DataFusion insert data into CSV and JSON tables. 

However, the User Experience / UX is tough as it is not easy to write datasets because: External tables require an existing location to exist, even if empty

So for example, if I wanted to write to json files in `/tmp/my_table` I currently need to create a the target file / directory externally

```shell
mkdir /tmp/my_table
echo "" > /tmp/my_table/1.json
datafusion-cli
```

```sql
❯ create external table my_table(x int, y int) stored as JSON location '/tmp/my_table';
0 rows in set. Query took 0.002 seconds.

❯ insert into my_table values (1,2), (3, 4);
+-------+
| count |
+-------+
| 2     |
+-------+
1 row in set. Query took 0.006 seconds.

❯
\q
```

```shell
$ cat /tmp/my_table/1.json

{"x":1,"y":2}
{"x":3,"y":4}
```

### Describe the solution you'd like

I would like to be able to have datafusion create the targets directly

Something like this, without any setup:

```sql
❯ create external table my_table(x int, y int) stored as JSON location '/tmp/my_table';
0 rows in set. Query took 0.002 seconds.

❯ insert into my_table values (1,2), (3, 4);
+-------+
| count |
+-------+
| 2     |
+-------+
1 row in set. Query took 0.006 seconds.
```
```shell
$ cat /tmp/my_table/1.json

{"x":1,"y":2}
{"x":3,"y":4}
```

## Option 1: Automatically create the target

One option is simply to remove the existence check on `CREATE EXTERNAL TABLE`

```sql
❯ create external table my_table(x int, y int) stored as JSON location '/tmp/my_table';
0 rows in set. Query took 0.002 seconds.

❯ insert into my_table values (1,2), (3, 4);
+-------+
| count |
+-------+
| 2     |
+-------+
1 row in set. Query took 0.006 seconds.
```

The downside is that the reason `CREATE EXTERNAL TABLE` errors if the target doesn't exist is to help people debug errors when reading

## Option 2: Add new DDL to `CREATE EXTERNAL TABLE`

Perhaps we could add a phrase to `CREATE EXTERNAL TABLE` like `FOR WRITE`

```sql
❯ create external table my_table(x int, y int) stored as JSON location '/tmp/my_table' FOR WRITE;
0 rows in set. Query took 0.002 seconds.
```

The semantics would be if the target file/directory doesn't already exist, then create it rather than error

## Option 3: Add a config parameter to control the default behavior

We could also add a config parameter like

```sql
❯ set catalog.auto_create_tables = true;
SET

❯ create external table my_table(x int, y int) stored as JSON location '/tmp/my_table' FOR WRITE;
0 rows in set. Query took 0.002 seconds.
```

with the same semantics as option 2 (create it target doesn't exist)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by devinjdangelo**
## Which issue does this PR close?

Follow on to #7244 

Related to #7228 and #7036

Note: does not close #7228 since it does not provide a mechanism to create external table target if it does not exist. It does offer a partial workaround since you can now specify `OPTIONS (insert_mode append_new_files)` when creating external tables so `insert into` will add new files to the table.

## Rationale for this change

Previous PRs adding support for inserting to `ListingTables` have not adequately tested an end to end process of creating an empty external table from SQL and subsequently inserting data into it. This PR adds test cases for this and adds a few enhancements / solves a few bugs to make the test cases pass.

## What changes are included in this PR?

- New test cases which create empty external table and insert data into it
- Remove logic throwing error when parquet schema is specified via SQL since it is now supported
- Support setting `ListingTableInsertMode` via `Create External Table ... OPTIONS ()` statement
- Update old test cases expecting setting parquet schema to fail
- Fix bug in CsvSink which set header to always be true even if table specified this explicitly as false

## Are these changes tested?

Yes

## Are there any user-facing changes?

Creating empty external tables and inserting to them via SQL is possible now.

### CODE REVIEW

File: datafusion/core/src/datasource/listing/table.rs [hunk 1]
```diff
@@ -2096,4 +2230,64 @@ mod tests {
...
+           .await?;
+
+        // check count
+        let batches = session_ctx
+            .sql("select * from foo")
+            .await?
+            .collect()
+            .await?;
+
+        let expected = vec![
```
Comment (alamb) [inline datafusion/core/src/datasource/listing/table.rs:2280]: this is very cool.

File: datafusion/core/src/datasource/listing/table.rs [hunk 2]
```diff
@@ -2096,4 +2230,64 @@ mod tests {
         // Return Ok if the function
         Ok(())
     }
+
+    /// tests insert into with end to end sql
+    /// create external table + insert into statements
+    async fn helper_test_insert_into_sql(
```
Comment (alamb) [inline datafusion/core/src/datasource/listing/table.rs:2236]: I think we could use sqllogictest to test this as well -- 

Perhaps add to https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/sqllogictests/test_files/insert.slt

I can't remember how temporary directories work in sqllogictests though

File: datafusion/sql/tests/sql_integration.rs [hunk 3]
```diff
@@ -1796,11 +1796,15 @@ fn create_external_table_with_compression_type() {
...
-    );
+    let expected = "CreateExternalTable: Bare { table: \"t\" }";
+    quick_test(sql, expected);
+}
+
+#[test]
+fn create_external_table_parquet_sort_order() {
+    let sql = "create external table foo(a varchar, b varchar, c timestamp) stored as parquet location '/tmp/foo' with order (c)";
+    let expected = "CreateExternalTable: Bare { table: \"foo\" }";
+    quick_test(sql, expected);
```
Comment (alamb) [inline datafusion/sql/tests/sql_integration.rs:1807]: 🤔  I tried to run this locally it doesn't quite work

```
❯ create external table t(x integer, y varchar) stored as parquet location '/tmp/foo' with order (x);
0 rows in set. Query took 0.001 seconds.

❯ select * from t;
0 rows in set. Query took 0.002 seconds.

❯ insert into t values (1, 'foo'), (2, 'bar');
This feature is not implemented: Writing to a sorted listing table via insert into is not supported yet. To write to this table in the meantime, register an equivalent table with file_sort_order = vec![]
```
But it is getting very close.

**Review by alamb — APPROVED** [review]
Thank you @devinjdangelo  -- this code is both nicely written and well tested and I think moves DataFusion forward. We are still not quite there but we are so close 

Thank you for making progress

File: datafusion/sql/tests/sql_integration.rs (hunk 3 above)
Comment (devinjdangelo) [reply datafusion/sql/tests/sql_integration.rs:1807]: Yes, that second error I added intentionally. Inserting to a sorted table would work, but there is nothing to enforce that the sort order is preserved yet. So, my concern is a user inserts unsorted data to a sorted table, and then subsequent queries return incorrect results surprisingly. 

We could add this as an issue to the streaming writes epic (support inserts to a sorted listingtable).

File: datafusion/core/src/datasource/listing/table.rs (hunk 2 above)
Comment (devinjdangelo) [reply datafusion/core/src/datasource/listing/table.rs:2236]: I actually worked on this in #7283 which I just opened for copy.slt. Once those updates are in, I can cut another PR to add additional tests to `insert.slt`

**Pull request merged**